[Bug]: Python PostCommit test suite is failing for BigQueryIO read tests

### What happened?

PostCommits are continuously failing with failures in `apache_beam/io/gcp/bigquery_read_it_test.py`

Recent run - https://github.com/apache/beam/actions/runs/7635974244/job/20802248794

```
FAILED apache_beam/io/gcp/bigquery_read_it_test.py::ReadTests::test_native_source - apache_beam.runners.dataflow.dataflow_runner.DataflowRuntimeException: Dataflow pipeline failed. State: FAILED, Error:
Traceback (most recent call last):
  File "apache_beam/runners/common.py", line 1435, in apache_beam.runners.common.DoFnRunner.process
  File "apache_beam/runners/common.py", line 851, in apache_beam.runners.common.PerWindowInvoker.invoke_process
  File "apache_beam/runners/common.py", line 997, in apache_beam.runners.common.PerWindowInvoker._invoke_process_per_window
  File "/runner/_work/beam/beam/sdks/python/apache_beam/transforms/core.py", line 1961, in <lambda>
    wrapper = lambda x, *args, **kwargs: [fn(x, *args, **kwargs)]
  File "/runner/_work/beam/beam/sdks/python/apache_beam/testing/util.py", line 192, in _equal
    raise BeamAssertException(msg)
apache_beam.testing.util.BeamAssertException: Failed assert: [{'number': 1, 'str': 'abc'}, {'number': 2, 'str': 'def'}, {'number': 3, 'str': '??????'}, {'number': 4, 'str': '????????????'}] == [{'number': 1, 'str': 'abc'}, {'number': 1, 'str': 'abc'}, {'number': 2, 'str': 'def'}, {'number': 2, 'str': 'def'}, {'number': 4, 'str': '????????????'}, {'number': 4, 'str': '????????????'}, {'number': 3, 'str': '??????'}, {'number': 3, 'str': '??????'}], unexpected elements [{'number': 1, 'str': 'abc'}, {'number': 2, 'str': 'def'}, {'number': 4, 'str': '????????????'}, {'number': 3, 'str': '??????'}]
```

### Issue Priority

Priority: 2 (default / most bugs should be filed as P2)

### Issue Components

- [X] Component: Python SDK
- [ ] Component: Java SDK
- [ ] Component: Go SDK
- [ ] Component: Typescript SDK
- [X] Component: IO connector
- [ ] Component: Beam YAML
- [ ] Component: Beam examples
- [ ] Component: Beam playground
- [ ] Component: Beam katas
- [ ] Component: Website
- [ ] Component: Spark Runner
- [ ] Component: Flink Runner
- [ ] Component: Samza Runner
- [ ] Component: Twister2 Runner
- [ ] Component: Hazelcast Jet Runner
- [ ] Component: Google Cloud Dataflow Runner

<img width="1072" alt="image" src="https://github.com/apache/beam/assets/25881114/047bc1aa-fa0c-4a58-94a1-e4dfdd957df6">
 is also failing with error:

 ```
 FAILED apache_beam/io/gcp/bigquery_file_loads_test.py::BigQueryFileLoadsIT::test_batch_copy_jobs_with_no_input_schema - AssertionError: 
Expected: (<[<BigqueryFullResultMatcher(Expected data is [(0,), (1,), (2,), (3,), (4,)])>, <BigqueryFullResultMatcher(Expected data is [(5,), (6,), (7,), (8,), (9,)])>]>)
     but: <[<BigqueryFullResultMatcher(Expected data is [(0,), (1,), (2,), (3,), (4,)])>, <BigqueryFullResultMatcher(Expected data is [(5,), (6,), (7,), (8,), (9,)])>]> was <DataflowPipelineResult 2024-01-23_22_23_28-9024953742164757457 DONE>
 ```

 which was recently added in #30015

CC: @ahmedabu98

Taking a look at `BigQueryFileLoadsIT.test_batch_copy_jobs_with_no_input_schema` now.

Although looking at the error coming out from `BigQueryFileLoadsIT. test_multiple_destinations_transform`, it looks like a separate issue:
```
E               "error": {
E                 "code": 400,
E                 "message": "Load configuration must specify at least one source URI",
E                 "errors": [
E                   {
E                     "message": "Load configuration must specify at least one source URI",
E                     "domain": "global",
E                     "reason": "invalid"
E                   }
E                 ],
E                 "status": "INVALID_ARGUMENT"
E               }
```

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -516,26 +516,26 @@ tasks.register("python38PostCommit") {
 
 tasks.register("python39PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py39:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py39:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py39:hdfsIntegrationTest")
-  dependsOn(":sdks:python:test-suites:portable:py39:postCommitPy39")
-  // TODO (https://github.com/apache/beam/issues/23966)
-  // Move this to Python 3.10 test suite once tfx-bsl has python 3.10 wheel.
-  dependsOn(":sdks:python:test-suites:direct:py39:inferencePostCommitIT")
+//  dependsOn(":sdks:python:test-suites:direct:py39:postCommitIT")
+//  dependsOn(":sdks:python:test-suites:direct:py39:hdfsIntegrationTest")
+//  dependsOn(":sdks:python:test-suites:portable:py39:postCommitPy39")
+//  // TODO (https://github.com/apache/beam/issues/23966)
+//  // Move this to Python 3.10 test suite once tfx-bsl has python 3.10 wheel.
+//  dependsOn(":sdks:python:test-suites:direct:py39:inferencePostCommitIT")
 }
 
 tasks.register("python310PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py310:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py310:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py310:hdfsIntegrationTest")
-  dependsOn(":sdks:python:test-suites:portable:py310:postCommitPy310")
+//  dependsOn(":sdks:python:test-suites:direct:py310:postCommitIT")
+//  dependsOn(":sdks:python:test-suites:direct:py310:hdfsIntegrationTest")
+//  dependsOn(":sdks:python:test-suites:portable:py310:postCommitPy310")
 }
 
 tasks.register("python311PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py311:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py311:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py311:hdfsIntegrationTest")
-  dependsOn(":sdks:python:test-suites:portable:py311:postCommitPy311")
+//  dependsOn(":sdks:python:test-suites:direct:py311:postCommitIT")
+//  dependsOn(":sdks:python:test-suites:direct:py311:hdfsIntegrationTest")
+//  dependsOn(":sdks:python:test-suites:portable:py311:postCommitPy311")
 }
 
 tasks.register("portablePythonPreCommit") {

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -503,39 +503,39 @@ tasks.register("pythonFormatterPreCommit") {
 
 tasks.register("python38PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py38:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py38:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py38:hdfsIntegrationTest")
-//  dependsOn(":sdks:python:test-suites:portable:py38:postCommitPy38")
+  dependsOn(":sdks:python:test-suites:direct:py38:postCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py38:hdfsIntegrationTest")
+  dependsOn(":sdks:python:test-suites:portable:py38:postCommitPy38")
   // TODO: https://github.com/apache/beam/issues/22651
   // The default container uses Python 3.8. The goal here is to
   // duild Docker images for TensorRT tests during run time for python versions
   // other than 3.8 and add these tests in other python postcommit suites.
-//  dependsOn(":sdks:python:test-suites:dataflow:py38:inferencePostCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py38:inferencePostCommitIT")
+  dependsOn(":sdks:python:test-suites:dataflow:py38:inferencePostCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py38:inferencePostCommitIT")
 }
 
 tasks.register("python39PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py39:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py39:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py39:hdfsIntegrationTest")
-//  dependsOn(":sdks:python:test-suites:portable:py39:postCommitPy39")
-//  // TODO (https://github.com/apache/beam/issues/23966)
-//  // Move this to Python 3.10 test suite once tfx-bsl has python 3.10 wheel.
-//  dependsOn(":sdks:python:test-suites:direct:py39:inferencePostCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py39:postCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py39:hdfsIntegrationTest")
+  dependsOn(":sdks:python:test-suites:portable:py39:postCommitPy39")
+  // TODO (https://github.com/apache/beam/issues/23966)
+  // Move this to Python 3.10 test suite once tfx-bsl has python 3.10 wheel.
+  dependsOn(":sdks:python:test-suites:direct:py39:inferencePostCommitIT")
 }
 
 tasks.register("python310PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py310:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py310:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py310:hdfsIntegrationTest")
-//  dependsOn(":sdks:python:test-suites:portable:py310:postCommitPy310")
+  dependsOn(":sdks:python:test-suites:direct:py310:postCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py310:hdfsIntegrationTest")
+  dependsOn(":sdks:python:test-suites:portable:py310:postCommitPy310")
 }
 
 tasks.register("python311PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py311:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py311:postCommitIT")
-//  dependsOn(":sdks:python:test-suites:direct:py311:hdfsIntegrationTest")
-//  dependsOn(":sdks:python:test-suites:portable:py311:postCommitPy311")
+  dependsOn(":sdks:python:test-suites:direct:py311:postCommitIT")
+  dependsOn(":sdks:python:test-suites:direct:py311:hdfsIntegrationTest")
+  dependsOn(":sdks:python:test-suites:portable:py311:postCommitPy311")
 }
 
 tasks.register("portablePythonPreCommit") {

--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -503,15 +503,15 @@ tasks.register("pythonFormatterPreCommit") {
 
 tasks.register("python38PostCommit") {
   dependsOn(":sdks:python:test-suites:dataflow:py38:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py38:postCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py38:hdfsIntegrationTest")
-  dependsOn(":sdks:python:test-suites:portable:py38:postCommitPy38")
+//  dependsOn(":sdks:python:test-suites:direct:py38:postCommitIT")
+//  dependsOn(":sdks:python:test-suites:direct:py38:hdfsIntegrationTest")
+//  dependsOn(":sdks:python:test-suites:portable:py38:postCommitPy38")
   // TODO: https://github.com/apache/beam/issues/22651
   // The default container uses Python 3.8. The goal here is to
   // duild Docker images for TensorRT tests during run time for python versions
   // other than 3.8 and add these tests in other python postcommit suites.
-  dependsOn(":sdks:python:test-suites:dataflow:py38:inferencePostCommitIT")
-  dependsOn(":sdks:python:test-suites:direct:py38:inferencePostCommitIT")
+//  dependsOn(":sdks:python:test-suites:dataflow:py38:inferencePostCommitIT")
+//  dependsOn(":sdks:python:test-suites:direct:py38:inferencePostCommitIT")
 }
 
 tasks.register("python39PostCommit") {

--- a/sdks/python/apache_beam/io/gcp/bigquery.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery.py
@@ -1854,6 +1854,7 @@ class WriteToBigQuery(PTransform):
       kms_key=None,
       batch_size=None,
       max_file_size=None,
+      max_partition_size=None,
       max_files_per_bundle=None,
       test_client=None,
       custom_gcs_temp_location=None,
@@ -1934,6 +1935,8 @@ bigquery_v2_messages.TableSchema`. or a `ValueProvider` that has a JSON string,
       max_file_size (int): The maximum size for a file to be written and then
         loaded into BigQuery. The default value is 4TB, which is 80% of the
         limit of 5TB for BigQuery to load any file.
+      max_partition_size (int): Maximum byte size for each load job to
+        BigQuery. Defaults to 15TB. Applicable to FILE_LOADS only.
       max_files_per_bundle(int): The maximum number of files to be concurrently
         written by a worker. The default here is 20. Larger values will allow
         writing to multiple destinations without having to reshard - but they
@@ -2059,6 +2062,7 @@ bigquery_v2_messages.TableSchema`. or a `ValueProvider` that has a JSON string,
     # TODO(pabloem): Consider handling ValueProvider for this location.
     self.custom_gcs_temp_location = custom_gcs_temp_location
     self.max_file_size = max_file_size
+    self.max_partition_size = max_partition_size
     self.max_files_per_bundle = max_files_per_bundle
     self.method = method or WriteToBigQuery.Method.DEFAULT
     self.triggering_frequency = triggering_frequency
@@ -2202,6 +2206,7 @@ bigquery_v2_messages.TableSchema`. or a `ValueProvider` that has a JSON string,
           with_auto_sharding=self.with_auto_sharding,
           temp_file_format=self._temp_file_format,
           max_file_size=self.max_file_size,
+          max_partition_size=self.max_partition_size,
           max_files_per_bundle=self.max_files_per_bundle,
           custom_gcs_temp_location=self.custom_gcs_temp_location,
           test_client=self.test_client,

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py
@@ -748,21 +748,17 @@ class TriggerLoadJobs(beam.DoFn):
     if not self.bq_io_metadata:
       self.bq_io_metadata = create_bigquery_io_metadata(self._step_name)
 
-    logging.critical("SOURCE URIS: %s", files)
-    try:
-      job_reference = self.bq_wrapper.perform_load_job(
-          destination=table_reference,
-          source_uris=files,
-          job_id=job_name,
-          schema=schema,
-          write_disposition=self.write_disposition,
-          create_disposition=create_disposition,
-          additional_load_parameters=additional_parameters,
-          source_format=self.source_format,
-          job_labels=self.bq_io_metadata.add_additional_bq_job_labels(),
-          load_job_project_id=self.load_job_project_id)
-    except Exception as e:
-      raise RuntimeError("SOURCE URIs: %s. GOT ERROR: %s", files, e)
+    job_reference = self.bq_wrapper.perform_load_job(
+        destination=table_reference,
+        source_uris=files,
+        job_id=job_name,
+        schema=schema,
+        write_disposition=self.write_disposition,
+        create_disposition=create_disposition,
+        additional_load_parameters=additional_parameters,
+        source_format=self.source_format,
+        job_labels=self.bq_io_metadata.add_additional_bq_job_labels(),
+        load_job_project_id=self.load_job_project_id)
     yield pvalue.TaggedOutput(
         TriggerLoadJobs.ONGOING_JOBS, (destination, job_reference))
     self.pending_jobs.append(

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py
@@ -747,6 +747,8 @@ class TriggerLoadJobs(beam.DoFn):
     )
     if not self.bq_io_metadata:
       self.bq_io_metadata = create_bigquery_io_metadata(self._step_name)
+
+    logging.critical("SOURCE URIS: %s", files)
     job_reference = self.bq_wrapper.perform_load_job(
         destination=table_reference,
         source_uris=files,

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads.py
@@ -749,17 +749,20 @@ class TriggerLoadJobs(beam.DoFn):
       self.bq_io_metadata = create_bigquery_io_metadata(self._step_name)
 
     logging.critical("SOURCE URIS: %s", files)
-    job_reference = self.bq_wrapper.perform_load_job(
-        destination=table_reference,
-        source_uris=files,
-        job_id=job_name,
-        schema=schema,
-        write_disposition=self.write_disposition,
-        create_disposition=create_disposition,
-        additional_load_parameters=additional_parameters,
-        source_format=self.source_format,
-        job_labels=self.bq_io_metadata.add_additional_bq_job_labels(),
-        load_job_project_id=self.load_job_project_id)
+    try:
+      job_reference = self.bq_wrapper.perform_load_job(
+          destination=table_reference,
+          source_uris=files,
+          job_id=job_name,
+          schema=schema,
+          write_disposition=self.write_disposition,
+          create_disposition=create_disposition,
+          additional_load_parameters=additional_parameters,
+          source_format=self.source_format,
+          job_labels=self.bq_io_metadata.add_additional_bq_job_labels(),
+          load_job_project_id=self.load_job_project_id)
+    except Exception as e:
+      raise RuntimeError("SOURCE URIs: %s. GOT ERROR: %s", files, e)
     yield pvalue.TaggedOutput(
         TriggerLoadJobs.ONGOING_JOBS, (destination, job_reference))
     self.pending_jobs.append(

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
@@ -925,9 +925,6 @@ class BigQueryFileLoadsIT(unittest.TestCase):
         create_disposition='CREATE_IF_NEEDED',
         write_disposition='WRITE_APPEND')
 
-    # reduce load job size to induce copy jobs
-    bqfl._DEFAULT_MAX_FILE_SIZE = 10
-    bqfl._MAXIMUM_LOAD_SIZE = 20
     verifiers = [
         BigqueryFullResultMatcher(
             project=self.project,
@@ -960,7 +957,10 @@ class BigQueryFileLoadsIT(unittest.TestCase):
           p | beam.Create(items) | bigquery.WriteToBigQuery(
               table=callable_table,
               create_disposition="CREATE_NEVER",
-              write_disposition="WRITE_APPEND"))
+              write_disposition="WRITE_APPEND",
+              # reduce load job size to induce copy jobs
+              max_file_size=10,
+              max_partition_size=20))
 
     hamcrest_assert(p, all_of(*verifiers))
 

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
@@ -903,7 +903,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
     _LOGGER.info(
         "Created dataset %s in project %s", self.dataset_id, self.project)
 
-  @pytest.mark.it_postcommit
+  @pytest.mark.it_postcommit_temp
   def test_batch_copy_jobs_with_no_input_schema(self):
     schema_1 = "col_1:INTEGER"
     schema_2 = "col_2:INTEGER"
@@ -949,8 +949,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
         dest += "_2"
       return dest
 
-    args = self.test_pipeline.get_full_options_as_args(
-        on_success_matcher=all_of(verifiers))
+    args = self.test_pipeline.get_full_options_as_args()
 
     with beam.Pipeline(argv=args) as p:
       # 0...4 going to table 1
@@ -965,7 +964,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
 
     hamcrest_assert(p, all_of(*verifiers))
 
-  @pytest.mark.it_postcommit
+  @pytest.mark.it_postcommit_temp
   def test_multiple_destinations_transform(self):
     output_table_1 = '%s%s' % (self.output_table, 1)
     output_table_2 = '%s%s' % (self.output_table, 2)

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
@@ -1000,8 +1000,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
                   if 'foundation' in d])
     ]
 
-    args = self.test_pipeline.get_full_options_as_args(
-        on_success_matcher=all_of(*pipeline_verifiers))
+    args = self.test_pipeline.get_full_options_as_args()
 
     with beam.Pipeline(argv=args) as p:
       input = p | beam.Create(_ELEMENTS, reshuffle=False)
@@ -1042,7 +1041,8 @@ class BigQueryFileLoadsIT(unittest.TestCase):
               create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
               write_disposition=beam.io.BigQueryDisposition.WRITE_EMPTY,
               max_file_size=20,
-              max_files_per_bundle=-1))
+              max_files_per_bundle=1))
+    hamcrest_assert(p, all_of(*pipeline_verifiers))
 
   @pytest.mark.it_postcommit
   def test_bqfl_streaming(self):

--- a/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
+++ b/sdks/python/apache_beam/io/gcp/bigquery_file_loads_test.py
@@ -903,7 +903,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
     _LOGGER.info(
         "Created dataset %s in project %s", self.dataset_id, self.project)
 
-  @pytest.mark.it_postcommit_temp
+  @pytest.mark.it_postcommit
   def test_batch_copy_jobs_with_no_input_schema(self):
     schema_1 = "col_1:INTEGER"
     schema_2 = "col_2:INTEGER"
@@ -964,7 +964,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
 
     hamcrest_assert(p, all_of(*verifiers))
 
-  @pytest.mark.it_postcommit_temp
+  @pytest.mark.it_postcommit
   def test_multiple_destinations_transform(self):
     output_table_1 = '%s%s' % (self.output_table, 1)
     output_table_2 = '%s%s' % (self.output_table, 2)
@@ -1041,7 +1041,7 @@ class BigQueryFileLoadsIT(unittest.TestCase):
               create_disposition=beam.io.BigQueryDisposition.CREATE_IF_NEEDED,
               write_disposition=beam.io.BigQueryDisposition.WRITE_EMPTY,
               max_file_size=20,
-              max_files_per_bundle=1))
+              max_files_per_bundle=-1))
     hamcrest_assert(p, all_of(*pipeline_verifiers))
 
   @pytest.mark.it_postcommit

--- a/sdks/python/test-suites/dataflow/common.gradle
+++ b/sdks/python/test-suites/dataflow/common.gradle
@@ -133,7 +133,7 @@ task postCommitIT {
         "test_opts": testOpts,
         "sdk_location": project.ext.sdkLocation,
         "suite": "postCommitIT-df${pythonVersionSuffix}",
-        "collect": "it_postcommit_temp",
+        "collect": "it_postcommit",
     ]
     def cmdArgs = mapToArgString(argMap)
     exec {

--- a/sdks/python/test-suites/dataflow/common.gradle
+++ b/sdks/python/test-suites/dataflow/common.gradle
@@ -133,7 +133,7 @@ task postCommitIT {
         "test_opts": testOpts,
         "sdk_location": project.ext.sdkLocation,
         "suite": "postCommitIT-df${pythonVersionSuffix}",
-        "collect": "it_postcommit"
+        "collect": "it_postcommit_temp",
     ]
     def cmdArgs = mapToArgString(argMap)
     exec {

